### ISSUE TITLE
Transformation Function in Data Iterator

### ISSUE THREAD

**Issue by ilkarman** [dmlc/mxnet#3054]
It doesn't seem possible to apply a transformation (e.g. dictionary lookup, encoding, etc) when evoking: mx.io.NDArrayIter()

However, I can find any documentation for wrapping my own generator (which applies the transformation) in a class that model.fit() will accept.

My way around is instead of using the 

model = mx.model.FeedForward()
model.fit()

approach, bind the symbol to the executor and then manually feed data and weights (because then I can use my custom function)

`cnn_exec = cnn.simple_bind(ctx=ctx, grad_req='add', **input_shapes)`

```
m = CNNModel(cnn_exec=cnn_exec,
             symbol=cnn,
             data=data, 
             label=label,
             param_blocks=param_blocks)
```

param_blocks is created like so:

```
param_blocks = []
arg_dict = dict(zip(arg_names, cnn_exec.arg_arrays))
for i, name in enumerate(arg_names):
    if name in ['softmax_label', 'data']:  # input, output
        continue
    initializer(name, arg_dict[name])
    param_blocks.append((i, arg_dict[name], args_grad[name], name))
```

The training:

```
# Stochastic gradient descent
optimizer = mx.optimizer.create(
    'sgd',
    learning_rate=LR,
    momentum = MOM,
    wd=WD
)
        
updater = mx.optimizer.get_updater(optimizer)
for iteration in range(EPOCHS):
    for batchX, batchY in manual_iterator(train_x, train_y, batch_size=100):
        m.data[:] = batchX
        m.label[:] = batchY
        # forward
        m.cnn_exec.forward(is_train=True)
        # backward
        m.cnn_exec.backward()
        # update weights
        for idx, weight, grad, name in m.param_blocks:
            updater(idx, grad, weight)
```

However, my model doesn't seem to be learning (the accuracy doesn't compare to the other approach). I'm also not sure if m.cnn_exec.backward() is to get the weights to update them or if it's some LSTM specific feature?

**Comment by piiswrong** [comment]
You can inherit mx.io.DataIter to create your own data iterator. You only need to implement provide_data, provide_label, next()
